line_follower.py

```diff
@@ -0,0 +1,218 @@
+    adopt_new_turn_multiplier = 0.98   # multiplier to determine when to switch the next direction to the drection
+    decrease_turn_multiplier = 0.8     # multiplier to determine if the value of the turn field should decrease
+    
+    def prepareImgRed(self, video):
+
+        retval, img = video.read()
+
+        #arrays to determine color masks
+        lower_red = np.array([0, 0, 30])
+        upper_red = np.array([70, 70, 255])
+        lower_blue = np.array([50, 0, 0])
+        upper_blue = np.array([255, 255, 255])
+
+        # apply masks
+        mask_red = cv2.inRange(img, lower_red, upper_red)
+        im_red = cv2.bitwise_and(img, img, mask=mask_red)
+        im_red = cv2.GaussianBlur(im_red,(5,5),0)
+        #threshold
+        ret_r, im_red = cv2.threshold(im_red, 60, 255, cv2.THRESH_BINARY)
+
+        # flatten so findContours doesn't get mad
+        im_red = cv2.cvtColor(im_red, cv2.COLOR_BGR2GRAY)
+
+        return img, im_red
+
+    def getDir(self):
+        return self.direction
+
+    def findStartDir(self, video):
+        img, im_red = self.prepareImgRed(video)
+        contours_r, hierarchy_r = cv2.findContours(im_red, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+        cnt = max(contours_r, key=cv2.contourArea)
+        hull = cv2.convexHull(cnt)
+        (x,y), (MA,ma), angle = cv2.fitEllipse(hull)
+
+        # determine if line is to the right or down based on the angle of the convex hull
+        if(angle < self.start_angle_barrier):
+            self.direction = Direction.down
+            self.next_dir = Direction.down
+        else:
+            self.direction = Direction.right
+            self.next_dir = Direction.right
+
+    def determineDir(self, video):
+        img, im_red = self.prepareImgRed(video)
+
+        # find contours
+        contours_r, hierarchy_r = cv2.findContours(im_red, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+        if len(contours_r) > 0:
+            
+            # compute largest contour, hull, and bounding rectangle
+            cnt = max(contours_r, key=cv2.contourArea)    
+            x, y, w, h = cv2.boundingRect(cnt)
+            hull = cv2.convexHull(cnt) 
+            boundingRect = cv2.minAreaRect(hull)
+            boundingRect = cv2.boxPoints(boundingRect)
+            boundingRect = np.int0(boundingRect)
+            hullArea = cv2.contourArea(hull)
+            current_turn = cv2.contourArea(boundingRect)/hullArea
+            print(self.direction)
+            if self.find_turn:     # looking for turn
+                 x_values = []
+                 y_values = []
+
+                 #determine if end of line reached based on contour solidity
+                 if (float(cv2.contourArea(cnt)) / hullArea) > self.end_of_line_solidity:
+                    self.is_moving = False
+                 #group x and y values based on distance from each other
+                 for x in hull[:,:,0]:
```

Comment:
this needs to be commented more and or pull out things to functions that are self documenting